File: 2021/day2/day2.js
const {readData} = require('../utils/utils')

const resolvePart1= (inputFile) => {
    const data = readData(inputFile)
    let x = 0, y = 0
    data.forEach(line => {
        line.startsWith('forward')
            ? x += Number(line.replace('forward ',''))
            : line.startsWith('up')
                ? y += Number(line.replace('up ',''))
                : y -= Number(line.replace('down ',''))
    })
    return Math.abs(x*y)
}

const resolvePart2 = (inputFile) => {
    const data = readData(inputFile)
    let horizontal = 0, depth = 0
    let aim = 0
    data.forEach(line => {
        if(line.startsWith('up')) depth -= Number(line.replace('up ',''))
        else if (line.startsWith('down')) depth += Number(line.replace('down ',''))
        else {
            const forward = Number(line.replace('forward ',''))
            horizontal += forward
            aim += forward * depth
        }

    })
    return Math.abs(horizontal*aim)
}

console.log(resolvePart1('day2_input_sample'))
console.log(resolvePart1('day2_input'))


console.log(resolvePart2('day2_input_sample'))
console.log(resolvePart2('day2_input'))

// Another way to read data

File: 2021/day2/day2.py
def get_data(filename):
    with open(filename) as file:
        lines = file.read().splitlines()
    return [(direction, int(n)) for direction, n in (line.split() for line in lines)]


def follow_course(course):
    horizontal_position = 0
    depth = 0
    aim = 0
    for direction, n in course:
        if direction == "forward":
            horizontal_position += n
            depth += aim * n
        elif direction == "up":
            aim -= n
        elif direction == "down":
            aim += n
    return horizontal_position, depth, aim


def part_1(horizontal_position, _, aim):
    return horizontal_position * aim


def part_2(horizontal_position, depth, _):
    return horizontal_position * depth


if __name__ == "__main__":
    sample_data = get_data('day2_input_sample')
    hda_sample = follow_course(sample_data)
    assert part_1(*hda_sample) == 150
    assert part_2(*hda_sample) == 900

    challenge_data = get_data('day2_input')
    hda_challenge = follow_course(challenge_data)
    print(part_1(*hda_challenge))
    print(part_2(*hda_challenge))